### PR TITLE
feat: add functions (sort, sort_by, max_by, min_by, map) to JMESPath visitor

### LINKED ISSUE #596: Implement full support for JMESPath expression types
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue, please leave a comment

## Describe the Feature

Implementing waiters support in #572 added support for the subset of JMESPath expression types needed to support existing AWS services but several expression types were left for future implementation. Specifically, the following need to be supported in codegen:

* ~and expressions~
* current node expressions
* expression type expressions
* functions expressions (besides `contains` and `length`, which are currently supported)
* index expressions
* literal expressions (besides string, number, boolean, and null, which are currently supported
* multiselect hash expressions
* ~not expressions~
* ~or expressions~
* slice expressions

See [KotlinJmespathExpressionVisitor.kt](https://github.com/awslabs/smithy-kotlin/blob/b79e1ef788fea8e8debfc816023b5cbc720a8ca3/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitor.kt) for a list of supported/unsupported expression types.

## Is your Feature Request related to a problem?

The lack of these expression types means that future waiters which use them will not codegen correctly.

## Proposed Solution

Implement support for the missing expression types.

## Describe alternatives you've considered

None.

## Additional Context

None.

- [x] I may be able to implement this feature request

### PR DESCRIPTION
**Pull request by 0marperez**
## Issue \#

closes #596 
## Description of changes

-Added sort to JMESPath visitor
-Added sort_by to JMESPath visitor
-Added max_by to JMESPath visitor
-Added min_by to JMESPath visitor
-Added map to JMESPath visitor

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitor.kt [hunk 1]
```diff
@@ -334,9 +344,36 @@ class KotlinJmespathExpressionVisitor(
...
 
+    private fun mappingFunction(
+        expression: FunctionExpression,
+        variableName: String,
+        function: (JmespathExpression, VisitedExpression) -> String,
+        invertedArgs: Boolean = false,
+    ): VisitedExpression {
+        val (argIndex, expressionIndex) = if (invertedArgs) 1 to 0 else 0 to 1
+
+        codegenReq(expression.arguments.size == 2) { "Unexpected number of arguments to $this" }
```
Comment (lauzadis) [inline codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitor.kt:372]: Let's print the number of expected arguments and the number of actual arguments received

File: codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitor.kt [hunk 2]
```diff
@@ -159,14 +169,16 @@ class KotlinJmespathExpressionVisitor(
                 append("if ($isNullExpr) null else ")
             }
 
-            val unSafeComparatorExpr = "compareTo(${right.identifier}) ${expression.comparator} 0"
+            val unSafeComparatorExpr = "compareTo(${right.identifier})"
```
[the remark this answers is not in the document]
Comment (aajtodd) [reply codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitor.kt:172]: Also this generates code like:

```kotlin
            val unSafeComparator = if (integer == null) null else integer?.compareTo(number)
            val safeComparator = if (unSafeComparator == null) null else unSafeComparator == 0
            safeComparator == true
```

Where you're just taking the null value if it's null otherwise comparing it to the expected value. You can compare `null` directly though

```kotlin
val comparator = integer == 0
```

Another example:
```kotlin
            val lists = it.lists
            val strings = lists?.strings
            val length = strings?.length ?: 0
            val number = 0.0
            val unSafeComparator = if (length == null) null else length.compareTo(number)
            val safeComparator = if (unSafeComparator == null) null else unSafeComparator > 0
            safeComparator == true
```

it can be condensed to:

```kotlin
            val safeComparator = length > 0
```

If the left expression of the comparision is nullable it can be changed to:

```
val safeComparator = length?.let { it > 0 }
```

Which will yield `Boolean?` as the result.

File: codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitor.kt [hunk 3]
```diff
@@ -334,9 +344,36 @@ class KotlinJmespathExpressionVisitor(
...
         else -> throw CodegenException("Unknown function type in $expression")
     }
 
+    private fun mappingFunction(
+        expression: FunctionExpression,
+        variableName: String,
+        function: (JmespathExpression, VisitedExpression) -> String,
+        invertedArgs: Boolean = false,
+    ): VisitedExpression {
+        val (argIndex, expressionIndex) = if (invertedArgs) 1 to 0 else 0 to 1
```
Comment (lauzadis) [inline codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitor.kt:370]: Can you explain why `map` has inverted arguments? I'm wondering if there is a "nicer" way to handle this?

File: codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitor.kt [hunk 4]
```diff
@@ -548,6 +585,37 @@ class KotlinJmespathExpressionVisitor(
...
+        operation: String,
+        stringOrNumberCheck: Boolean = true,
+    ): String {
+        val argName = arg.identifier
+        val result = bestTempVarName(resultName)
+        writer.withBlock("val $result = $argName?.$operation{", "}") {
+            val expressionValue = addTempVar("expression", subfieldCodegen((expression as ExpressionTypeExpression).expression as FieldExpression, "it"))
+            if (stringOrNumberCheck) {
+                write("if ($expressionValue as Any !is Number && $expressionValue as Any !is String) throw Exception(\"Result of applying expression should be string or number\")")
+            }
```
Comment (lauzadis) [inline codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitor.kt:601]: Same comment about `invertedArgs` here, why is this needed for `map` only?

File: codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitor.kt [hunk 5]
```diff
@@ -548,6 +585,37 @@ class KotlinJmespathExpressionVisitor(
...
+    private fun mapFunctionLogic(
+        expression: JmespathExpression,
+        arg: VisitedExpression,
+        resultName: String,
+        operation: String,
+        stringOrNumberCheck: Boolean = true,
+    ): String {
+        val argName = arg.identifier
+        val result = bestTempVarName(resultName)
+        writer.withBlock("val $result = $argName?.$operation{", "}") {
```
Comment (lauzadis) [inline codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitor.kt:597]: nit: space between `$operation` and `{`

File: codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitor.kt [hunk 6]
```diff
@@ -548,6 +585,37 @@ class KotlinJmespathExpressionVisitor(
...
+            val expressionValue = addTempVar("expression", subfieldCodegen((expression as ExpressionTypeExpression).expression as FieldExpression, "it"))
+            if (stringOrNumberCheck) {
+                write("if ($expressionValue as Any !is Number && $expressionValue as Any !is String) throw Exception(\"Result of applying expression should be string or number\")")
```
Comment (lauzadis) [inline codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitor.kt:600]: Can we throw a more descriptive exception type than the base `Exception`?

File: runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/util/JMESPath.kt [hunk 7]
```diff
@@ -72,3 +72,27 @@ public fun Any?.type(): String = when (this) {
...
+@JvmName("FloatJmespathSort")
+public fun List<Float>.jmespathSort(): List<Float> = this.sorted()
+
+@InternalApi
+@JvmName("LongJmespathSort")
+public fun List<Long>.jmespathSort(): List<Long> = this.sorted()
+
+@InternalApi
+@JvmName("DoubleJmespathSort")
+public fun List<Double>.jmespathSort(): List<Double> = this.sorted()
```
Comment (lauzadis) [inline runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/util/JMESPath.kt:98]: Can this be made generic? You would lose the `@JvmName` annotation but also remove redundant functions.
If the annotation is not required, I'd try to make these generic for any `T`

File: codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitor.kt [hunk 8]
```diff
@@ -548,6 +585,37 @@ class KotlinJmespathExpressionVisitor(
         return notNull
     }
 
+    private fun mapFunctionLogic(
```
Comment (lauzadis) [inline codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitor.kt:588]: I'd call this `applyFunction`, since `mapFunctionLogic` could be confused for the actual `map` function, and I think the `Logic` suffix is unnecessary

Comment (lauzadis) [reply codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitor.kt:588]: Would it make sense to make this an extension function of `JmespathExpression`? Then you can use it like this

```kt
private fun JmespathExpression.applyFunction(
   ...
) { ... }

expression.applyFunction(arg, "sorted", "sortedBy")
```

File: codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitor.kt (hunk 6 above)
Comment (aajtodd) [inline codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitor.kt:600]: **correctness**:

This check is wrong:

```kotlin
            val sorted = structs?.sortedBy{
                val expression = it?.integer
                if (expression as Any !is Number && expression as Any !is String) throw Exception("Result of applying expression should be string or number")
                expression
            }
```

`expression` is of type `Any?` so `as Any` is incorrect. Expression is likely never going to be null anyway though so I guess in practice it will be `Any` at runtime. If we can we should generate the code as `it.integer` not `it?` here since `.sortedBy` will only run if `structs` isn't null in the first place (due to `?`). 

Backing up for a second, why do we even need to generate the `stringOrNumber` check in the first place? I see in the [JMESPath spec](https://jmespath.org/specification.html#sort-by):

> If the result of evaluating the expr against the current array element results in type other than a number or a string, an invalid-type error will occur.

But why do we need to worry about this? Seems like this would be an invalid model most likely.

File: codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitor.kt (hunk 5 above)
Comment (aajtodd) [inline codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitor.kt:597]: **nit**: add space between `$operation` and opening brace `{`

File: codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitor.kt (hunk 4 above)
Comment (0marperez) [reply codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitor.kt:601]: The [map](https://jmespath.org/specification.html#map) expression can result in any type but for the rest ([max_by](https://jmespath.org/specification.html#max-by), `min_by`, etc..) the expression should result in a string or a number.

File: codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitor.kt (hunk 6 above)
Comment (0marperez) [reply codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitor.kt:600]: > But why do we need to worry about this? Seems like this would be an invalid model most likely.

The expression resulting in anything other than a string or number is not being checked for at the model level. I think it makes sense to have that check. Someone could add an expression that results in something else and it would fail without an exception explaining why

File: runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/util/JMESPath.kt [hunk 9]
```diff
@@ -72,3 +72,27 @@ public fun Any?.type(): String = when (this) {
     null -> "null"
     else -> throw Exception("Undetected type for: $this")
 }
+
+@InternalApi
+@JvmName("StringJmespathSort")
+public fun List<String>.jmespathSort(): List<String> = this.sorted()
```
[the remark this answers is not in the document]
Comment (0marperez) [reply runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/util/JMESPath.kt:78]: I added them because JMESPath's `sort` can only be applied to numbers and strings. I think it makes sense to check the type and then throw an exception explaining that if it's the wrong type of list

File: codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitor.kt (hunk 6 above)
Comment (aajtodd) [reply codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitor.kt:600]: Also Smithy is doing some validation on the JMESPath expressions in the model: https://github.com/smithy-lang/smithy/blob/main/smithy-jmespath/src/main/java/software/amazon/smithy/jmespath/TypeChecker.java

I don't think we need to add runtime validations here, this is the kind of stuff that belongs as an error in the model such that it never can even get into the hands of a customer.

File: runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/util/JMESPath.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (aajtodd) [reply runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/util/JMESPath.kt:78]: Again I don't think we want to be responsible for runtime checking things that could be caught by model validation. These just inflate the size of the runtime core artifact with no real value. I think we should remove them and just call sorted directly which is already defined with generics

**Review by aajtodd — APPROVED** [review]

**Review by ianbotsf — APPROVED** [review]
